### PR TITLE
Firearms rework again [Modular Attachments]

### PR DESCRIPTION
**Pull request by ynot01**
# Better sprites needed
# Document the changes in your pull request

## Gun changes

Every gun now has slight variance when firing to justify the presence of accuracy-improving attachments.

Energy weapons now have very tiny recoil, mostly for flavor.

Ballistic weapons now have a little more recoil but can still largely be ignored.

Demonstration of variance and shake:
demonstration of variance and shake

## Attachments

Creates a framework for universal firearm attachments.

This system is independent of the flashlight & bayonet attachments, they coexist currently but in the future they can be merged.

Guns can specify what attachments can fit on it through the whitelist `available_attachments` list variable.

Guns can also specify how many total attachments they can carry through the `max_attachments` variable.

Currently all energy & ballistic weapons can carry the 4 current attachments.

Laser sights utilize similar code to beam rifles (particle acceleration rifles) and can be toggled or changed color through buttons.

buttons

laser sight line

See wiki to see what every other attachment presently does.

# Spriting
full of codersprites

Laser sights by OrcaCora

# Wiki Documentation

Simple sights reduce recoil by 3

Holo sights reduce recoil by 6

Laser sights reduce recoil by 5 when turned on

Vertical grip reduces recoil by 0.5

Holo/laser/vert can be bought for 2TC
All can be printed at sec protolathe

# Changelog

:cl:  ynot01, MechanoidVoid
rscadd: Added simple sights to security protolathe, locked under sec tech
rscadd: Added holographic sights to security protolathe, locked under adv. weapon tech
rscadd: Added vertical grips to security protolathe, locked under weapon tech
rscadd: Added laser sights to security protolathe, locked under weapon tech
tweak: All non-magical guns now have innate 5 spread (Does not stack with weapons that already had spread)
tweak: Ballistic weapons now have innate 0.3 recoil (Does not stack with weapons that already had recoil)
tweak: Energy weapons now have innate 0.1 recoil (Does not stack with weapons that already had recoil)
/:cl:

### CODE REVIEW

File: code/datums/action.dm [hunk 1]
```diff
@@ -204,6 +204,61 @@
...
+	if(att)
+		att.toggle_on()
+	UpdateButtonIcon()
+
+/datum/action/item_action/toggle_laser_sight/UpdateButtonIcon(status_only = FALSE, force)
+	if(att)
+		if(att.is_on)
+			button_icon_state = "laser_sight_on"
+		else
+			button_icon_state = "laser_sight"
```
Comment (tattax) [inline code/datums/action.dm]:
```suggestion
/datum/action/item_action/toggle_laser_sight/UpdateButtonIcon(status_only = FALSE, force)
	button_icon_state = "laser_sight[att?.is_on ? "_on" : ""]"
```

File: code/datums/action.dm [hunk 2]
```diff
@@ -204,6 +204,61 @@
...
+/datum/action/item_action/toggle_laser_sight/Trigger()
+	if(!att)
+		if(istype(target, /obj/item/gun))
+			var/obj/item/gun/H = target
+			for(var/obj/item/attachment/A in H.current_attachments)
+				if(istype(A, /obj/item/attachment/laser_sight))
+					att = A
+					break
+	if(att)
+		att.toggle_on()
```
Comment (tattax) [inline code/datums/action.dm]:
```suggestion
	att?.toggle_on()
```
this is what conditionals are for, not using many useless statements that make things harder to read

File: code/datums/action.dm [hunk 3]
```diff
@@ -204,6 +204,61 @@
...
+	name = "Toggle Laser Sight"
+	icon_icon = 'icons/obj/guns/attachment.dmi'
+	button_icon_state = "laser_sight"
+	var/obj/item/attachment/laser_sight/att
+
+/datum/action/item_action/toggle_laser_sight/Trigger()
+	if(!att)
+		if(istype(target, /obj/item/gun))
+			var/obj/item/gun/H = target
+			for(var/obj/item/attachment/A in H.current_attachments)
```
Comment (tattax) [inline code/datums/action.dm]:
```suggestion
			var/obj/item/gun/parent_gun = target
			for(var/obj/item/attachment/A in parent_gun.current_attachments)
```
H for gun when you have G at your disposal or any other word is kinda too much

**Comment by 00ze-cyclone** [comment]
So you added a slightly nauseating effect when shooting guns without any attachment and added to the uplink items crew can print at a ridiculous TC cost.
Also laser sights, I guess for when you want to tell your target exactly where you will shoot if you can’t remember where you want to aim.
Can you explain how it make gunfights more interesting in any way?

File: code/datums/action.dm [hunk 4]
```diff
@@ -204,6 +204,61 @@
...
+			return
+		att.laser_color = C
+	UpdateButtonIcon()
+
+/datum/action/item_action/change_laser_sight_color/UpdateButtonIcon(status_only = FALSE, force)
+	if(att)
+		if(att.is_on)
+			button_icon_state = "laser_sight_on"
+		else
+			button_icon_state = "laser_sight"
```
Comment (tattax) [inline code/datums/action.dm]:
```suggestion
	button_icon_state = "laser_sight[att?.is_on ? "_on" : ""]"
```

File: code/modules/projectiles/attachments/_attachment.dm [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	var/attachment_type = 0
+
+	/// "You slide the attachment into place on gun."
+	var/attach_verb = "slide"
+
+	var/mob/current_user = null
+
+	/// List of actions to add to the gun when attached.
+	/// See code/modules/projectiles/attachments/laser_sight.dm for example.
+	var/actions_list = list()
```
Comment (tattax) [inline code/modules/projectiles/attachments/_attachment.dm]:
```suggestion
	var/list/actions_list = list()
```

File: code/modules/projectiles/attachments/_attachment.dm [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+/obj/item/attachment/update_icon()
+	if(is_on)
+		icon_state = "[initial(icon_state)]_on"
+	else
+		icon_state = initial(icon_state)
+	. = ..()
+	if(attached_gun)
+		attached_gun.update_attachments()
```
Comment (tattax) [inline code/modules/projectiles/attachments/_attachment.dm]:
```suggestion
	icon_state = initial(icon_state) + is_on ? "_on" : ""
	. = ..()
	attached_gun?.update_attachments()
```

File: code/modules/projectiles/attachments/_attachment.dm [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	
+	if(G.loc == user)
+		set_user(user)
+	G.attachment_flags |= attachment_type
+	G.current_attachments += src
+	G.update_attachments()
+	forceMove(G)
+
+	if(user)
+		current_user = user
```
Comment (tattax) [inline code/modules/projectiles/attachments/_attachment.dm]:
```suggestion
/obj/item/attachment/proc/on_attach(obj/item/gun/G, mob/user = null)
	attached_gun = G

	for(var/act in actions_list)
		var/datum/action/attachment_action = new act(G)
		G.attachment_actions += attachment_action
		if(user && G.loc == user)
			L.Grant(user)
	
	if(G.loc == user)
		set_user(user)
	G.attachment_flags |= attachment_type
	G.current_attachments += src
	G.update_attachments()
	forceMove(G)

	if(user)
		current_user = user
```

File: code/modules/projectiles/attachments/_attachment.dm [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+/// Called when the attachment is detached from a weapon
+/obj/item/attachment/proc/on_detach(obj/item/gun/G, mob/living/user = null)
+	for(var/act in actions_list)
+		for(var/S in G.attachment_actions)
+			if(istype(S, act))
+				G.attachment_actions -= S
+				var/datum/action/A = S
+				A.Remove(user)
```
Comment (tattax) [inline code/modules/projectiles/attachments/_attachment.dm]: could use some better names

File: code/modules/projectiles/attachments/laser_sight.dm [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	if(!check_user())
+		// Doesn't need to be optimized, runs nothing if the list is already empty
+		QDEL_LIST(attached_gun.current_tracers)
+		return
+	process_aim()
+	var/diff = abs(aiming_lastangle - lastangle)
+	if(diff < 0.1 && !force_update)
+		return
+	aiming_lastangle = lastangle
+	var/obj/item/projectile/beam/beam_rifle/hitscan/aiming_beam/P = new
```
Comment (tattax) [inline code/modules/projectiles/attachments/laser_sight.dm:70]: why p

File: code/modules/projectiles/attachments/laser_sight.dm [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+/obj/item/attachment/laser_sight/proc/delay_penalty(amount)
+	aiming_time_left = clamp(aiming_time_left + amount, 0, aiming_time)
+
+/obj/item/attachment/laser_sight/Destroy()
+	set_user(null)
+	listeningTo = null
+	STOP_PROCESSING(SSfastprocess, src)
+	if(attached_gun)
+		QDEL_LIST(attached_gun.current_tracers)
```
Comment (tattax) [inline code/modules/projectiles/attachments/laser_sight.dm]:
```suggestion
	QDEL_LIST(attached_gun?.current_tracers)
```

File: code/modules/projectiles/gun.dm [hunk 11]
```diff
@@ -169,7 +189,7 @@
 
 
 /obj/item/gun/proc/shoot_live_shot(mob/living/user, pointblank = 0, atom/pbtarget = null, message = 1)
-	if(recoil)
+	if(recoil > 0)
```
Comment (tattax) [inline code/modules/projectiles/gun.dm:192]: the current should do this without putting the extra >

File: code/modules/projectiles/gun.dm [hunk 12]
```diff
@@ -532,13 +608,39 @@
...
+		var/mutable_appearance/M = mutable_appearance('icons/obj/guns/attachment.dmi', "[A.icon_state]_a")
+		M.pixel_x = att_position * 6
+		add_overlay(M, TRUE)
+		attachment_overlays += M
+		att_position += 1
+
+	update_icon(TRUE)
+	for(var/X in actions)
+		var/datum/action/A = X
+		A.UpdateButtonIcon()
```
Comment (tattax) [inline code/modules/projectiles/gun.dm]:
```suggestion
	for(var/datum/action/A as anything in actions)
		A.UpdateButtonIcon()
```

File: code/modules/projectiles/gun.dm (hunk 11 above)
Comment (ynot01) [reply code/modules/projectiles/gun.dm:192]: negative values are truthy, we do not want it triggering on negative values

File: code/datums/action.dm (hunk 2 above)
Comment (ynot01) [reply code/datums/action.dm]: don't have to be rude about it

File: code/modules/projectiles/attachments/laser_sight.dm (hunk 9 above)
Comment (ynot01) [reply code/modules/projectiles/attachments/laser_sight.dm:70]: P for projectile, copied from beam rifle code

**Comment by ynot01** [comment]
> So you added a slightly nauseating effect when shooting guns without any attachment

I'd say even "slightly nauseating" is an overstatement.

> and added to the uplink items crew can print at a ridiculous TC cost.

Adding it to the uplink was a courtesy because I didn't want this to be a strictly security buff. On certain syndicate weapons these attachments can be quite powerful.

> Also laser sights, I guess for when you want to tell your target exactly where you will shoot if you can’t remember where you want to aim. Can you explain how it make gunfights more interesting in any way?

The aiming laser is both an upside and a downside. It provides 5 bonus accuracy on a non-exclusive gun part (under/sidebarrel), so the cost of that is opponents can see you coming or tell where you are aiming with the laser.

I feel this makes fights a bit more interactive as you can use quick reactions to dodge shots, as well as for the person firing, it introduces another layer of potential mindgames. The original purpose of this, however, was just to line up shots and occupy a bit less thinking power on that front. So yes, I do believe this makes gunfights more interesting on all fronts.

**Comment by Rdrazga** [comment]
Ngl I do agree on the items being in the traitor menu being weird that they are the exact same as sec printable ones, (syndie better versions maybe?) also prob a bit overpriced for the same reason.

**Comment by SomeguyManperson** [comment]
Tc are too densified for it to not be overpriced without being part of a bundle

**Comment by Marmio64** [comment]
not sure why weak energy guns that are pretty much just firing light, should have recoil. Only attachment I see that is interesting enough to keep would be the laser sight.

**Comment by Skrem7** [comment]
Like the idea, would 100% support getting it in to then get some testing done to see what could be adjusted better.

TC cost for attachments is high, agree with Theos. Should either be better, be bundled, be craftable when you purchase a gun, or any mix of the bunch.

For the love of god do not powercreep makeshift crafting even more

File: code/datums/action.dm (hunk 2 above)
Comment (tattax) [reply code/datums/action.dm]: just explaining it, maybe came out the wrong way

File: code/modules/projectiles/guns/ballistic.dm [hunk 13]
```diff
@@ -110,6 +110,15 @@
 	var/feedback_recoil_reverse = FALSE // TRUE for clockwise , FALSE for anti-clockwise
 	var/feedback_slide_close_move = TRUE // does the slide closing cause the gun to twist clockwise?
 
+	available_attachments = list(
+		/obj/item/attachment/scope/simple,
```
Comment (SomeguyManperson) [inline code/modules/projectiles/guns/ballistic.dm:114]: Is there a reason this isn't compacted into /scope for both of them?

Comment (ynot01) [reply code/modules/projectiles/guns/ballistic.dm:114]: readability? not every scope is guaranteed to fit on every rifle? not allowing the non-functional sight base to be mounted?

**Comment by Hopekz** [comment]
Decent PR imo because it adds a cool framework that other people can use in the future (:

**Review by JamieD1 — APPROVED** [review]

**Comment by Hopekz** [comment]
Now we see how laggy that laser beam is when like 2+ people turn it on

**Comment by Lichton** [comment]
I don't like my screen going insane when I pew pew, legit its visual clutter that throws off my aim a ton cus click hard

**Pull request merged**